src/main/telemetry/msp_shared.c

```diff
@@ -24,230 +24,294 @@
-    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+    sbufWriteU8(&responsePacket.buf, error);
+    responsePacket.result = TELEMETRY_MSP_RES_ERROR;
+    sbufSwitchToReader(&responsePacket.buf, msp_serial_outBuf);
 }
 
-bool handleMspFrame(uint8_t *frameStart, int frameLength, uint8_t *skipsBeforeResponse)
+//despite its name, the function actually handles CRSF frame payload with MSP in it
+//it reads the MSP into requestPacket stucture and handles it after receiving all the chunks.
+bool handleMspFrame(uint8_t *crsfPayload, uint8_t crsfPayloadLength, uint8_t *skipsBeforeResponse)
 {
+    if (crsfPayloadLength < MIN_LENGTH_CHUNK)
+        return false;   //prevent analyzing garbage data
+
     static uint8_t mspStarted = 0;
     static uint8_t lastSeq = 0;
 
-    if (sbufBytesRemaining(&mspPackage.responsePacket->buf) > 0) {
-        mspStarted = 0;
-    }
-
-    if (mspStarted == 0) {
-        initSharedMsp();
-    }
+    sbuf_t sbufInput;
 
-    mspPacket_t *packet = mspPackage.requestPacket;
-    sbuf_t *frameBuf = sbufInit(&mspPackage.requestFrame, frameStart, frameStart + (uint8_t)frameLength);
-    sbuf_t *rxBuf = &mspPackage.requestPacket->buf;
-    const uint8_t header = sbufReadU8(frameBuf);
-    const uint8_t seqNumber = header & TELEMETRY_MSP_SEQ_MASK;
-    const uint8_t version = (header & TELEMETRY_MSP_VER_MASK) >> TELEMETRY_MSP_VER_SHIFT;
+    const uint8_t status = crsfPayload[MSP_INDEX_STATUS];
+    const uint8_t seqNumber = status & crsf_msp_status_sequence_mask;
+    const uint8_t version = (status & crsf_msp_status_version_mask) >> crsf_msp_status_version_shift;
 
-    if (version != TELEMETRY_MSP_VERSION) {
+    if (version > TELEMETRY_MSP_VERSION) {
         sendMspErrorResponse(TELEMETRY_MSP_VER_MISMATCH, 0);
         return true;
     }
 
-    if (header & TELEMETRY_MSP_START_FLAG) {
-        // first packet in sequence
-        uint8_t mspPayloadSize = sbufReadU8(frameBuf);
-
-        packet->cmd = sbufReadU8(frameBuf);
-        packet->result = 0;
-        packet->buf.ptr = mspPackage.requestBuffer;
-        packet->buf.end = mspPackage.requestBuffer + mspPayloadSize;
-
-        checksum = mspPayloadSize ^ packet->cmd;
-        mspStarted = 1;
-    } else if (!mspStarted) {
-        // no start packet yet, throw this one away
-        return false;
-    } else if (((lastSeq + 1) & TELEMETRY_MSP_SEQ_MASK) != seqNumber) {
-        // packet loss detected!
-        mspStarted = 0;
-        return false;
+    if (status & crsf_msp_status_start_mask) { // first packet in sequence
+        uint16_t mspPayloadSize;
+        if(version == 1) {//MSPv1
+            if (crsfPayloadLength < MIN_LENGTH_REQUEST_V1)
+                return false;   //prevent analyzing garbage data
+
+            mspPayloadSize = crsfPayload[MSP_INDEX_SIZE_V1];
+            requestPacket.cmd = crsfPayload[MSP_INDEX_ID_V1];
+            if (0xff == mspPayloadSize) {//jumbo frame
```

Comment:
```mspPayloadSize == 0xff```

Reply:
Why? Is it in the coding style?
Putting rvalue before "==" is a safe indication that this is a comparison and not an assignment. If somebody puts '=' instead of '==' here, the code become non compileable. So, it's safer this way. 
Why putting even single 'return;' line in braces under if but not placing lvalue at the first place under 'if's?

Reply:
I don't mind changing but please explain why "**if (mspPayloadSize == 0xff)**" is better than "**if (0xff == mspPayloadSize)**". It's the same paranoia as with put everything in {}. ;-)

Reply:
Preference 🙂 Leave it like that if you prefer it, but as I mentioned in my comment at the bottom I think we should remove support for MSP v1 and jumbo frames. MSP v2 can do it all. 
Use of braces is in the Betaflight coding style https://github.com/betaflight/betaflight/blob/master/docs/development/CodingStyle.md